src/encrypted_container.rs

```diff
@@ -0,0 +1,248 @@
+/// including the 28 bytes of cryptographic data overhead, and leaving 11 bytes
+/// as the CBOR serialization overhead.
+///
+/// Decryption operation is done the same way as its counterpart, but backwards.
+/// The serialized Encrypted Data Container in bytes is first deserialized, making a EDC instance,
+/// and afterwards the decryption operation in Trussed is called, resulting in a original serialized
+/// object, which is then deserialized to a proper instance.
+///
+/// CBOR was chosen as the serialization format due to its simplicity and extensibility.
+/// If that is a requirement, more space efficient and faster would be postcard. Be advised however,
+/// that it's format changes between major revisions (as expected with semver versioning).
+///
+/// This type has implemented bidirectional serialization to trussed Message object.
+///
+/// Showing the processing paths graphically:
+/// T -> \[u8\]: object -> CBOR serialization -> EncryptedDataContainer encryption  -> CBOR serialization -> serialized EncryptedDataContainer
+///
+/// \[u8\] -> T: serialized EncryptedDataContainer -> CBOR deserialization -> EncryptedDataContainer decryption -> CBOR deserialization -> object
+///
+/// Note: to decrease the CBOR overhead it might be useful to rename the serialized object fields for
+/// the serialization purposes. Use the `#[serde(rename = "A")]` attribute.
+///
+/// Usage example:
+/// ```compile_fail
+/// let ek = get_encryption_key(trussed)?;
+/// let data = EncryptedDataContainer::from_obj(trussed, obj, None, ek)?;
+/// let data_serialized: Message = data.try_into()?;
+/// ```
+/// Future work and extensions:
+/// - Generalize over serialization method
+/// - Generalize buffer size
+/// - Investigate postcard structure extensibility, as a means for smaller overhead for serialization
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+pub struct EncryptedDataContainer {
+    #[serde(rename = "D")]
+    data: trussed::types::Message,
+    #[serde(rename = "T")]
+    tag: trussed::types::ShortData,
+    #[serde(rename = "N")]
+    nonce: trussed::types::ShortData,
+}
+
+impl TryFrom<&[u8]> for EncryptedDataContainer {
+    type Error = Error;
+
+    /// Create an instance from this serialized Encrypted Data Container
+    fn try_from(value: &[u8]) -> Result<Self> {
+        let credential =
+            cbor_deserialize(value).map_err(|_| Error::DeserializationToContainerError)?;
+        Ok(credential)
+    }
+}
+
+impl TryFrom<EncryptedDataContainer> for Message {
+    type Error = Error;
+
+    /// Try to serialize EncryptedDataContainer to Bytes
+    fn try_from(value: EncryptedDataContainer) -> Result<Self> {
+        let mut buf = [0u8; SERIALIZED_OBJECT_BUFFER_SIZE];
+        let r = cbor_serialize(&value, &mut buf).map_err(|_| Error::ObjectSerializationError)?;
+        let bytes = Message::from_slice(r).map_err(|_| Error::SerializationBufferTooSmall)?;
+        Ok(bytes)
```

Comment:
Serializing directly into a `Message` would safe us a buffer and a copy:

```
Message::try_from(|buf| {
    cbor_serialize(&value, &mut buf)
        .map(|s| s.len())
        .map_err(|_| Error::ObjectSerializationError))
})
```

Reply:
I was convinced this is a consuming operation. After checking the docs apparently there is nothing like that in the `heapless`, which would pack an existing slice as a new vec, which is a shame.
Anyway, good point! Indeed this is additional copy, worth removing. Done as suggested.